debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java

```diff
@@ -0,0 +1,337 @@
+            queue.offer(event);
+        }
+    }
+
+    public static <TResult> BufferingChangeStreamCursor<TResult> fromIterable(
+                                                                              ChangeStreamIterable<TResult> stream,
+                                                                              MongoDbTaskContext taskContext,
+                                                                              MongoDbStreamingChangeEventSourceMetrics metrics,
+                                                                              Clock clock) {
+        var config = taskContext.getConnectorConfig();
+
+        return new BufferingChangeStreamCursor<>(
+                new EventFetcher<>(stream, config.getMaxBatchSize(), metrics, clock, config.getPollInterval()),
+                Threads.newFixedThreadPool(MongoDbConnector.class, taskContext.serverName(), "replicator-buffer", 1),
+                config.getPollInterval());
+    }
+
+    /**
+     * Creates new prefetching cursor
+     *
+     * @param fetcher MongoDB change event fetcher
+     * @param executor executor used to dispatch buffering thread
+     * @param throttler throttling mechanism
+     */
+    public BufferingChangeStreamCursor(EventFetcher<TResult> fetcher, ExecutorService executor, DelayStrategy throttler) {
+        this.fetcher = fetcher;
+        this.executor = executor;
+        this.throttler = throttler;
+    }
+
+    public BufferingChangeStreamCursor(EventFetcher<TResult> fetcher, ExecutorService executor, Duration throttleMaxSleep) {
+        this(fetcher, executor, DelayStrategy.boundedExponential(Duration.ofMillis(1), throttleMaxSleep, 2));
+    }
+
+    public BufferingChangeStreamCursor<TResult> start() {
+        executor.submit(fetcher);
+        return this;
+    }
+
+    @Override
+    public ResumableChangeStreamEvent<TResult> tryNext() {
+        var event = pollWithDelay();
+        if (event != null) {
+            lastResumeToken = event.resumeToken;
+        }
+        return event;
+    }
+
+    /**
+     * Returns next event in buffer.
+     * Not that unlike other Mongo implementation this method does not block
+     *
+     * @throws NoSuchElementException when no element is available
+     * @return event
+     */
+    @Override
+    public ResumableChangeStreamEvent<TResult> next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+        return tryNext();
+    }
+
+    /**
+     * Repeatedly polling fetcher for new event, exponentially waiting between polls until limit is reached
+     *
+     * @return event or null if not available within time limit
+     */
+    private ResumableChangeStreamEvent<TResult> pollWithDelay() {
+        boolean slept;
+        ResumableChangeStreamEvent<TResult> event;
+
+        do {
+            event = fetcher.poll();
+            slept = throttler.sleepWhen(event == null);
+        } while (slept);
+
+        return event;
+    }
```

Comment:
This seems to work nicely in combination with the exponential delay strategy -- which I took the librety of slightly modifying (added the ability to also reset the time when the maximum delay is reached).

Reply:
It is questionable whether we should use this also in the fetcher loop. 

The same approach in the fetcher would actually increase the total sleep time. Since we would exponentially sleep after each empty polls. Considering there is a already an implicit timeout on the Mongo side I think the time sleeping might actually increase quite a bit. 

We could also combine the current approach with exponential delay and instead of flat delay every 5 empty polls we would engage the exponential delay. So every 5 empty events we would progressively wait longer until the max delay is reached and then reset. While that increases the number of cursor polls I think it would still provide sufficient throttling -- which would be at the same time lower compared to the one in the prefetching cursor, which is exactly what we want (as we don't want the fetcher to throttle more than reads from the buffer queue)

Reply:
@jcechace Sunds good